iree/turbine/kernel/wave/decompose_vmma_ops.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2024 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from ..wave.constraints import (
+    Constraint,
+    HardwareConstraint,
+    MMAType,
+    MMAOperand,
+)
+from iree.turbine.kernel.wave.utils import get_mfma_load_elems_per_thread
+from .._support.tracing import CapturedTrace
+from .._support.indexing import IndexSequence, IndexSymbol, IndexExpr
+from ..ops.wave_ops import (
+    get_custom,
+    MMA,
+    Reshape,
+)
+from ..lang.global_symbols import *
+
+import copy
+import sympy
+
+
+VMMA_TO_NATIVE_MAP = {
+    MMAType.F32_16x16x32_K8_F16: MMAType.F32_16x16x16_F16,
+    MMAType.F32_32x32x16_K8_F16: MMAType.F32_32x32x8_F16,
+}
+
+
+def get_m_dim(mma_op):
+    m_dims = set(mma_op.lhs_type.symbolic_shape).difference(
+        mma_op.rhs_type.symbolic_shape
+    )
+    assert len(m_dims) == 1, "Expect to have single M-dim."
+    return m_dims.pop()
+
+
+def get_n_dim(mma_op):
+    n_dims = set(mma_op.rhs_type.symbolic_shape).difference(
+        mma_op.lhs_type.symbolic_shape
+    )
+    assert len(n_dims) == 1, "Expect to have single N-dim."
+    return n_dims.pop()
+
+
+def get_k_dim(mma_op):
+    k_dims = set(mma_op.lhs_type.symbolic_shape).intersection(
+        mma_op.rhs_type.symbolic_shape
+    )
+    k_dims = k_dims.difference(mma_op.acc_type.symbolic_shape)
+    assert len(k_dims) == 1, "Expect to have single K-dim."
+    return k_dims.pop()
+
+
+def replace_subexpr(
+    main_expr: sympy.Expr, src_subexpr: sympy.Expr, dst_subexpr: sympy.Expr
+):
+    assert (
+        len(main_expr.find(src_subexpr)) > 0
+    ), "Expected src subexpression to be found in main expression."
+    # Early exit if src and dst is already the same.
+    if src_subexpr == dst_subexpr:
+        return
+    main_expr = main_expr.subs(src_subexpr, dst_subexpr)
+
+
+def decompose_vmma_ops(
+    trace: CapturedTrace,
+    constraints: list[Constraint],
+):
+    """
+    The decomposition for VMMA is done in five steps:
+      1. Collect all MMA ops and only process the ones with virtual intrinsics.
+      2. For each VirtualMMAOp, we generate check it's unroll factor
+      3. Generate slices of lhs and rhs from coaleseced reads based on native and virtual MMA sizes
+      4. Generate native MMA ops based on unroll factor that takes in slices of coaleseced lhs and rhs
+      5. Modify index expression of new mma op by replacing the expression based on virtual layout
+         with it's equivalent native layout expression.
+    """
+    # Get MMA nodes.
+    mma_nodes = trace.walk(lambda node: isinstance(get_custom(node), MMA))
+    if not mma_nodes:
+        return
+    hardware_constraint = next(
+        c for c in constraints if isinstance(c, HardwareConstraint)
+    )
+    for node in mma_nodes:
+        mma_op = get_custom(node)
+        mma_type = mma_op.mma_type
```

Comment:
how about
```
mma_type = mma_op.mma_type or hardware_constraint.mma_type
```
which gives you what you want because of python short circuiting?

Reply:
make sense, done, thanks!